src/background/services/wallet.ts

```diff
@@ -0,0 +1,332 @@
+      );
+    }
+
+    const exchangeRate = exchangeRates.rates[walletAddress.assetCode];
+    const getRateOfPay = (rate: string) =>
+      _getRateOfPay({
+        rate,
+        exchangeRate,
+        assetScale: walletAddress.assetScale,
+      });
+    rateOfPay = getRateOfPay(DEFAULT_RATE_OF_PAY);
+    minRateOfPay = getRateOfPay(MIN_RATE_OF_PAY);
+    maxRateOfPay = getRateOfPay(MAX_RATE_OF_PAY);
+
+    await this.openPaymentsService.initClient(walletAddress.id);
+    this.setConnectState('connecting');
+    this.events.emit('wallet.close_popup');
+
+    const [existingTab] = await this.browser.tabs.query({
+      url: this.browser.runtime.getURL(APP_URL),
+    });
+    try {
+      await this.outgoingPaymentGrantService.completeOutgoingPaymentGrant(
+        amount,
+        walletAddress,
+        recurring,
+        InteractionIntent.CONNECT,
+        existingTab?.id,
+      );
+    } catch (error) {
+      if (
+        isErrorWithKey(error) &&
+        error.key === 'connectWallet_error_invalidClient' &&
+        autoKeyAdd
+      ) {
+        if (!KeyAutoAddService.supports(walletAddress)) {
+          this.setConnectStateError(error);
+          throw new ErrorWithKey(
+            'connectWalletKeyService_error_notImplemented',
+          );
+        }
+        if (!autoKeyAddConsent) {
+          this.setConnectStateError(error);
+          throw new ErrorWithKey('connectWalletKeyService_error_noConsent');
+        }
+
+        // add key to wallet and try again
+        try {
+          const tabId =
+            await this.outgoingPaymentGrantService.addPublicKeyToWallet(
+              walletAddress,
+              existingTab?.id,
+            );
+          this.setConnectState('connecting');
+          await this.outgoingPaymentGrantService.completeOutgoingPaymentGrant(
+            amount,
+            walletAddress,
+            recurring,
+            InteractionIntent.CONNECT,
+            tabId,
+          );
+        } catch (error) {
+          this.setConnectStateError(error);
+          throw error;
+        }
+      } else {
+        this.setConnectStateError(error);
+        throw error;
+      }
+    }
+
+    await this.storage.set({
+      walletAddress,
+      rateOfPay,
+      minRateOfPay,
+      maxRateOfPay,
+      connected: true,
+    });
+    this.setConnectState(null);
+  }
+
+  async reconnectWallet({ autoKeyAddConsent }: ReconnectWalletPayload) {
+    if (!autoKeyAddConsent) {
+      await this.validateReconnect();
+      return;
+    }
+
+    const { walletAddress } = await this.storage.get(['walletAddress']);
+    if (!walletAddress) {
+      throw new Error('reconnectWallet_error_walletAddressMissing');
+    }
+    if (!KeyAutoAddService.supports(walletAddress)) {
+      throw new ErrorWithKey('connectWalletKeyService_error_notImplemented');
+    }
+    this.setConnectState('connecting');
+    this.events.emit('wallet.close_popup');
```

Comment:
This should be moved into the catch below - close only when adding public key, not before validation.